model/src/main/kotlin/PackageReference.kt

```diff
@@ -72,16 +73,26 @@ data class PackageReference(
     fun collectDependencies(
         maxDepth: Int = -1,
         filterPredicate: (PackageReference) -> Boolean = { true }
-    ): Set<Identifier> =
-        dependencies.fold(sortedSetOf()) { refs, ref ->
-            refs.also {
-                if (maxDepth != 0) {
-                    if (filterPredicate(ref)) it += ref.id
-                    it += ref.collectDependencies(maxDepth - 1, filterPredicate)
-                }
+    ): Set<Identifier> {
+        val result = mutableSetOf<Identifier>()
+
+        val queue = Stack<Pair<PackageReference, Int>>()
+        queue.push(Pair(this, 0))
+
+        while (queue.isNotEmpty()) {
+            val (pkg, level) = queue.pop()
+
+            if (filterPredicate(pkg)) result.add(pkg.id)
+
+            val dependenciesLevel = level + 1
```

Comment:
It seems arbitrary to call one of them `level` and the other `dependenciesLevel`, since both are dependency levels. How about `currentLevel` and `nextLevel` for example?

Reply:
Done.